src/performance.js

```diff
@@ -60,6 +68,49 @@ export class Performance {
     this.events_ = [];
   }
 
+  /**
+   * Listens to viewer and resource events.
+   */
+  coreServicesAvailable() {
+    this.viewer_ = viewerFor(this.win);
+    this.resources_ = resourcesFor(this.win);
+
+    this.viewer_.onVisibilityChanged(this.flush.bind(this));
+
+    let timeReqToBeVisible = 0;
+
+    // If doc didnt start in prerender mode, then no need to
+    // listen for when the first resource is laid out.
+    if (this.viewer_.didStartInPrerender()) {
+      this.resources_.whenFirstResourceLaidOut().then(() => {
+        if (this.viewer_.hasBeenVisible()) {
+          // Prerender Completed, and user has requested
+          // doc to be visible prior to prerender completion.
+          // User might have been seeing the first set of resources
+          // loading.
+          this.tick('pc', null, timer.now() - timeReqToBeVisible);
```

Comment:
@erwinmombay I just ran through it with @cramforce but please ping him to confirm if anything is still not clear. From what I now understand:

1. `hasBeenVisible` is the key. You can still keep `didStartInPrerender_`, but I don't really see much need for it in this PR.
2. `whenFirstResourceLaidOut` promise is not what we want. What we want here is to collect a set of elements that have to be loaded for the first pass: prerender or proper render. This is basically all elements that fall in `viewportRect` in the first pass through `discoverWork`. The only nuance - we need to consult other filters in the `scheduleLayoutOrPreload_` such as `prerenderAllowed`. It's possible we want to factor those filters outside into the `discoverWork` instead. Once they are all complete - this is our "visual completeness" signal.

Now, our main value to calculate is this:

```
let visibleTime = viewer.hasBeenVisible() ? now() : -1;
if (visibleTime == -1) {
  viewer.whenVisible().then(() => visibleTime = now());
}

let visualCompletenessTime = -1;
resources.whenFirstViewportLoaded().then(() => visualCompletenessTime = now());
```

Then, the number we want to report is:
```
const timeToVisualCompleteness = max(0, visualCompletenessTime - visibleTime);
```

It will be negative and thus 0 when all loading is done before the doc became visible. And it will be a positive number if loading is done after the document became visible.